WIP: start l10n tooling

groovecoder opened the pull request.

pdehaan commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,7 @@
+
+terms = Terms&nbsp;and&nbsp;Privacy
+
+copyright = Portions of this content are © 1998-2018 by individual mozilla.org contributors. Content available under a Creative Commons license
```
Actual markup below is:

```html
<p data-l10n-id="copyright">
  Portions of this content are &copy; 1998-2018 by individual mozilla.org contributors.
  <br />
  Content available under a <a {{> analytics/link_event eventLabel="Creative Commons"}} href="https://www.mozilla.org/foundation/licensing/website-content/" target="_blank" rel="noopener">Creative Commons license</a>.
</p>
```

Not sure if the .ftl files should use &amp;copy; entity or if we need anything special for link preserving.

pdehaan commented on public/locales/de/app.ftl:
```diff
@@ -0,0 +1,7 @@
+
+terms = Allgemeine&nbsp;Geschäftsbedingungen&nbsp;und&nbsp;Datenschutzerklärung
+
+copyright = Teile dieses Inhalts sind © 1998-2018 von einzelnen mozilla.org-Mitwirkenden. Inhalt verfügbar unter einer Creative Commons-Lizenz
```
/cc @flodolo @stasm

Do either of you have any guidance or suggestions on bundling the .ftl files directly in the source repo, versus doing something like a separate "blurts-server-l10n" repo for translations?

flodolo replied:
Reasons to use a different repository:
* History: Pontoon will write in the repository every 20 minutes, once for each locale (if there are changes). History is going to become messy very quickly.
* Decouplement: You can develop on master and decide when strings are ready for translation.

Cons:
* You need scripts to import strings back.
* You need to expose strings in the l10n repository via PR.

Alternatively, we can also write in a branch instead of master, but then you need to deal with merges back and forth.

Side note: we'll need a dev server updated frequently for l10n testing. For that it might be easier to have the strings in the main repository.



From the point of view of configuration, if you want Pontoon to write here we just need to add its bot.

groovecoder replied:
@flodolo can you point us to examples of each kind of setup? I.e., same-repo vs. separate-repo?

pdehaan replied:
The one separate-repo I can remember off the top of my head is FxA's content server:

https://github.com/mozilla/fxa-content-server
vs
https://github.com/mozilla/fxa-content-server-l10n

I think the trick was putting a package.json in the [*-l10n repo](https://github.com/mozilla/fxa-content-server-l10n/blob/master/package.json) so you can include it as a dependency in your main repo's package.json. But of course, as I went to prove that, I notice that FxA team is using some Bash script[1] to download locales, so we may want to sync up with them to see why they switched to doing it that way (or maybe I'm misremembering how it was done originally).

[1] https://github.com/mozilla/fxa-content-server/blob/master/scripts/download_l10n.sh

groovecoder commented on .env-dist:
```diff
@@ -4,6 +4,9 @@ SERVER_URL=http://localhost:6060
 COOKIE_SECRET=3895d33b5f9730f5eb2a2067fe0a690e298f55f5e382c032fd3656863412
 
+# available languages for l20n
+AVAILABLE_LANGUAGES="de,en,es,fr,it,ja,pt,ru,zh"
```
We should thoroughly test, but the `accept-language` module says that `get` will "Get the most likely language given an Accept-Language string".

groovecoder commented on middleware.js:
```diff
@@ -1,9 +1,21 @@
+
+function pickLanguage (req, res, next) {
+  acceptLanguage.languages(AppConstants.AVAILABLE_LANGUAGES.split(","));
+  req.app.locals.locale = acceptLanguage.get(req.headers["accept-language"]);
```
Code looks like it will default to whatever is the first language tag specified in `acceptLanguage.languages`. So in our case - yes.

stasm commented:
Thanks for the CC and for kicking this off.

`l20n.js` hasn't been actively developed in more than a year. We shifted focus completely to Fluent. I should really get to deprecating it officially (https://github.com/l20n/spec/issues/17).

Instead, we have developed [`fluent`](https://github.com/projectfluent/fluent.js/tree/master/fluent) (for parsing FTL files and formatting translations) and [`fluent-dom`](https://github.com/projectfluent/fluent.js/tree/master/fluent-dom) (for localizing the DOM), although they might require some additional work to set up. `fluent-dom` doesn't initialize the localization of the document automatically during the load of the page. Some glue code is needed to parse the links to FTL files form `<head>` and start the localization of the document.

There's an experimental packaged called [`fluent-web`](https://github.com/projectfluent/fluent-web) which provides this glue code together with the dependencies. We haven't had a strong driver to work on it recently. @zbraniecki will know the latest status of it.

For language negotiation there's also [`fluent-langneg`](https://github.com/projectfluent/fluent.js/tree/master/fluent-langneg) which is well-tested and used by a number of other Mozilla web projects. It might make sense to use it here, too, although other packages available on `npm` should be OK as well.

groovecoder commented:
I looked thru send, screenshots, testpilot, and fluent-web and I prefer fluent-web by a large margin. I tried to simply replace `l20n.js` with `fluent-web.js` but it looks like `fluent-web` has some making & building steps that I can't figure out.

How long would it take to get `fluent-web` up to a usable state like `l20n.js`? Can it be done by the end of this week?

If not, I'll start digging into using `fluent-dom` directly.

zbraniecki commented:
> but it looks like fluent-web has some making & building steps that I can't figure out.

```
hg clone https://github.com/projectfluent/fluent-web
cd fluent-web
npm i
make build
open ./examples/simple.html
```

does that work?

groovecoder commented:
I get `SyntaxError: import declarations may only appear at top level of a module fluent-web.js:2`

groovecoder commented:
I'm happy to move this over to a fluent-web issue.

groovecoder commented:
Still waiting to see if we can use `fluent-web` for the front-end l10n, but I made a significant change to this PR for back-end l10n ...

1. On server start-up:
   * scan `public/locales` directory for available languages
   * instantiate a `FluentBundle` for each language
   * keep both in `app.locals`
   * `app.use(pickLanguage)` middleware
2. During request:
   * `pickLanguage` checks `Accept-Language` and sets appropriate `req.locale` and `req.fluentBundle`
3. In back-end code, e.g., `controllers/home.js`, use `req.fluentBundle.format(req.fluentBundle.getMessage("notFound")` where it used to have the hard-coded English message.
   * (could probably condense this with a `locale.bundleMessage()` helper function

This seems to work fine!

<img width="832" alt="screen shot 2018-10-03 at 4 34 46 pm" src="https://user-images.githubusercontent.com/71928/46440844-47165580-c72a-11e8-90c8-3a707916273a.png">

And it feels simpler than some of the other approaches I saw in send, testpilot, screenshots, etc.

Does it seem sane?

pdehaan commented:
One further wrench into the plan is that the presence of a directory/file isn't always an indicator of translation %%.

In other repos (like [**mozilla/send**](https://github.com/mozilla/send/blob/master/scripts/get-prod-locales.js)) we had a script which would only include locales that are 100% translated. I think [**mozilla/fxa-content-server**](https://github.com/mozilla/fxa-content-server/blob/d95118004500ecd522061b4de08ace068e21ae37/grunttasks/l10n-supported-locales.js) had a lower threshold (90%) before it would be considered ready for production.

groovecoder commented on views/layouts/default.hbs:
```diff
@@ -44,9 +52,9 @@
+          <li data-l10n-id="support">Support</li>
+          <li><a data-l10n-id="feedback" {{> analytics/link_event eventLabel="Give Feedback"}} href="https://qsurvey.mozilla.com/s3/Firefox-Monitor-Feedback" target="_blank" rel="noopener">Give Feedback</a></li>
+          <li><a data-l10n-id="terms" {{> analytics/link_event eventLabel="Terms & Privacy"}} href="https://www.mozilla.org/privacy/firefox-monitor/{{> analytics/utm_site_footer }}" target="_blank" rel="noopener">Terms&nbsp;and&nbsp;Privacy</a></li>
           <li><a {{> analytics/link_event eventLabel="GitHub"}} href="https://github.com/mozilla/blurts-server/" target="_blank" rel="noopener noreferrer"><img src="/img/github.svg" alt="GitHub Logo" /></a></li>
```
They are! Good catch!

groovecoder commented:
Addressed all comments except the template-level ones. (Still checking if we'll be able to use `fluent-web` for template strings, or if we'll make a handlebars helper for them.)

Consolidated some of the server start-up code into a single `locale.loadLanguagesIntoApp` function.

Made the `pickLanguage` middleware also add a `fluentFormat` helper on the request object, so calling code can be simplified to, e.g., `req.fluentFormat("notFound")`

pdehaan commented:
> The %-checking heuristic seems neat, but how helpful is it?

I think it's helpful, but you'd probably need to sync up with the FxA team or Send/dcoates and see how well it's working for them after all this time.

> The fxa-content-server check just seems to check that a translated message is non-blank, which ignores up-to-date-ness of the translations; the send check for 100% no errors nor missing translations means (like you say in the PR) that any update to en-US invalidates all translations until they are back to 100%?

IIRC, I think Send stores all it's production locale list in package.json and it's manually managed.

https://github.com/mozilla/send/blob/a2eee15a7de297092575f28da13a0c330e64058e/package.json#L134-L185

The [script](https://github.com/mozilla/send/blob/master/scripts/get-prod-locales.js) just runs a scanner to say which locales are 100% or whatever it does, and you can either manually update package.json if you like the suggestions of adding the new translations, or there may have been some `--write` flag mischief that would auto-write the package.json changes for you.

> Either tooling tactic seems to rely on a significant amount of manual translation support anyway. Does the tooling help notify the human translators of the updates needed?

I don't think the tooling does anything (at least in the case of Send). I _assume_ it's just something that gets run before a deploy to see if there are any new locales that have achieved 90-100% translations and would be good candidates to include in the website.

I imagine we could write some unit test that gets triggered on CI that could do something in a more automated way if you think it'd be useful to be notified regularly of 100% translations. But I imagine that'd just be a bunch of failing PRs due to unrelated strings. Plus, I don't know how actionable updating `app.locals.AVAILABLE_LANGUAGES` is versus just doing it prior to a stage/prod deploy.

pdehaan commented on locale-utils.js:
```diff
@@ -0,0 +1,69 @@
+const LocaleUtils = {
+  init() {
+    const languageDirectories = fs.readdirSync( localesDir ).filter(item => {
+      return (item[0] !== "." && item.length === 2)
```
What does `item.length === 2` do?

It seems like it's checking if the file/folder name is 2 characters long.

pdehaan commented on middleware.js:
```diff
@@ -29,14 +64,18 @@ function clientErrorHandler (err, req, res, next) {
 
 
 function errorHandler (err, req, res, next) {
+  console.log("errorHandler, err.message: ", err.message);
```
`console.log()` or should we use **mozlog** here?

groovecoder replied:
Good catch. We shouldn't actually need this `console.log` since `log.error` in `logErrors` above should take care of the logging.

pdehaan commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,35 @@
+user-add-title = {-product-name} : Confirm Email
+
+user-verify-token-error = Verification token is required.
+user-verify-email-report-subject = Your Firefox Monitor report
```
nit: instead of "Firefox Monitor" we should probably use `{-product-name}` here.

pdehaan commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,35 @@
+layout-support = Support
+layout-feedback = Give Feedback
+layout-terms = Terms&nbsp;and&nbsp;Privacy
+layout-copyright = Portions of this content are &copy; 1998-2018 by individual mozilla.org contributors. <br />Content available under a <a {{> analytics/link_event eventLabel="Creative Commons"}} href="https://www.mozilla.org/foundation/licensing/website-content/" target="_blank" rel="noopener">Creative Commons license</a>.
```
Not sure how well fluent supports nested HTML, but I'm pretty sure it isn't great from a localizer standpoint since it can get pretty complex (for example, localizers will have to ignore the handlebars partial and that `eventLabel` which shouldn't be localized). I think l20n had ways to avoid this, but i don't know about fluent.

zbraniecki replied:
@flodolo ^ ?

pdehaan commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,35 @@
+
+error-not-subscribed = This email address is not subscribed to {-product-name}.
+error-hibp-throttled = Too many connections to HIBP.
+error-hibp-connect = Error connecting to HIBP.
```
Not sure how visible these errors are, or if we should use a more verbose "Have I Been Pwned" vs the abbrev.

flodolo replied:
In any case, HIBP needs an explanation in a comment.

flodolo replied:
Side note: looks like there's a term for HIBP now, it should be used instead.

flodolo commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,35 @@
+error-hibp-connect = Error connecting to HIBP.
+error-hibp-load-breaches = Could not load breaches.
+
+hibp-notify-email-subject = {-product-name} Alert : Your account was involved in a breach.
```
Spaces before a colon don't seem correct in English (also other strings)?

flodolo commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,35 @@
+user-verify-email-report-subject = Your Firefox Monitor report
+user-verify-title = {-product-name} : Subscribed
+
+user-unsubscribe-token-error = Unsubscribe requires a token.
```
Unsubscribing?

groovecoder commented:
@stasm @flodolo @pdehaan I'd like to merge & deploy this with English-only at first, just to avoid the bit-rot of a PR that touches almost every file. 😉 Does that seem sane?

flodolo commented:
> @stasm @flodolo @pdehaan I'd like to merge & deploy this with English-only at first, just to avoid the bit-rot of a PR that touches almost every file. 😉 Does that seem sane?

Works for me. We have space to fix the en-US until we start exposing for localization in Pontoon.

Where is German coming from btw?

groovecoder commented:
Oh, I was just dumping some of the strings thru Google Translate to make sure the messages were showing up correctly in all the different contexts. (e.g., English in error logs but German on the user-facing error page)

groovecoder commented:
Also fixed tests!

groovecoder commented:
@lesleyjanenorton is going to work on the handlebars templates tomorrow. Then we'll remove the temporary `de/app.ftl` file we used for testing, and try to merge & deploy this.

pdehaan commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,35 @@
+
+layout-support = Support
+layout-feedback = Give Feedback
+layout-terms = Terms&nbsp;and&nbsp;Privacy
```
Nit: we should probably see if we can do this using something like [`white-space`](https://css-tricks.com/almanac/properties/w/whitespace/) in CSS.

stasm replied:
Something like the following should do the trick:

```properties
layout-terms = Terms{"\u00A0"}and{"\u00A0"}Privacy
```

stasm commented on hbs-helpers.js:
```diff
@@ -16,7 +23,7 @@ function breachDataClasses(dataClasses) {
   const jsDate = new Date(date);
-  // TODO: localize this
+  // TODO: l10n
   return jsDate.toLocaleDateString("en-US", {year: "numeric", month: "long", day: "numeric"});
```
Fluent can format dates, too. You can pass the date object as an argument to `format`. I assume you'd need to make `prettyDate` take `req` as an argument, however. I'm not sure if that's possible given where it's used.

```js
function prettyDate(req, date) {
  return fluentFormat(req, "pretty-date", {date: new Date(date)});
}
```
```properties
# en-US
pretty-date = {DATETIME($date, year: "numeric", month: "long", day: "numeric")}
```
```properties
# Some other language might decide long month names are too long.
pretty-date = {DATETIME($date, year: "numeric", month: "short", day: "numeric")}
```

There's also a not-very-well-documented feature of setting the default formatting options for the arguments.

```js
import {FluentDateTime} from "fluent";

function prettyDate(req, date) {
  return fluentFormat(req, "pretty-date", {
    date: new FluentDateTime(date, {
      year: "numeric",
      month: "long",
      day: "numeric",
    })
  });
}
```
```properties
# en-US
pretty-date = {$date}
```

Other locales can still customize the formatting options, if needed:

```properties
# Some other language might decide long month names are too long.
pretty-date = {DATETIME($date, year: "numeric", month: "short", day: "numeric")}
```

stasm commented on middleware.js:
```diff
@@ -1,9 +1,35 @@
+  req.pickedLanguage = pickedLanguage;
+  const fluentBundle = req.app.locals.FLUENT_BUNDLES[pickedLanguage];
+  req.fluentFormat = (message, args = null, errors = null) => {
+    return fluentBundle.format(fluentBundle.getMessage(message, args));
```
This code will work fine in the optimistic scenario of most localizations being complete :) To better support the more realistic scenario of some localizations missing some translations, I'd like to suggest basing this code on the concept of _languages as a list_.

The idea is to negotiate the _list of languages_ requested by the user against the list of available languages. The output is another list. The default language should always be the last item of the resulting list. You can use the `negotiateLanguages` function from the `fluent-langneg` package to perform this negotation reliably.

Unfortunately, AFAICT, the `accept-language` package doesn't expose a method which returns a list of languages parsed from the header. `fluent-langneg` provides a very simple implementation which should be enough for this use-case.

Here's how I'd use it:

```js
import {negotiateLanguages, acceptedLanguages} from "fluent-langneg";

function pickLanguage (req, res, next) {
  // Parse the header into a list of language tags.
  const requested = acceptedLanguages(req.headers["accept-language"]);
  // Negotiate the list of languages for the current user, in order of their preference.
  const supportedLocales = negotiateLanguages(
    requested,
    req.app.locals.AVAILABLE_LANGUAGES
    {defaultLocale: "en-US"}
  );

  req.fluentFormat = (id, args = null, errors = null) => {
    for (let locale of supportedLocales) {
      let bundle = req.app.locals.FLUENT_BUNDLES[locale];
      if (bundle.hasMessage(id)) {
        const message = bundle.getMessage(id);
        return bundle.format(message, args);
      }
    }
    // If no bundle has this message (even the default),
    // fall back to the identifier.
    return id;
  };

  next();
}
```

stasm replied:
If you don't go this route (although I think you should), note that there's a small bug in the current code. The `args` object should be passed as the second argument to `format` rather than `getMessage`. After all, a single message might be used to produce many different translations in the UI, each for different arguments.

```diff
-return fluentBundle.format(fluentBundle.getMessage(message, args));
+return fluentBundle.format(fluentBundle.getMessage(message), args);
```

stasm commented on locale-utils.js:
```diff
@@ -0,0 +1,65 @@
+      Error.captureStackTrace(this, FluentError);
+    }
+
+    this.fluentID = fluentID;
```
👍 This is a very good pattern, thanks for doing it this way!

pdehaan commented:
Once we merge this, it'd also be nice to set up a **CODEOWNERS** file and then make @flodolo a required reviewer of any `public/locales/en/*.ftl` string changes (unless he doesn't want to, in which case, don't).

A good reference is how we did it in [**mozilla-services/screenshots** CODEOWNERS](https://github.com/mozilla-services/screenshots/blob/master/CODEOWNERS).

pdehaan commented:
Re: 
> _"is there tooling around `CODEOWNERS` files? I.e., can we enforce that **flodolo** has to approve PRs that contain changes to `public/locales/en/app.ftl`?"_

https://help.github.com/articles/about-codeowners/ says:

> Code owners are automatically requested for review when someone opens a pull request that modifies code that they own. When someone with admin or owner permissions has enabled required reviews, they also can optionally require approval from a code owner before the author can merge a pull request in the repository.

groovecoder commented:
@pdehaan brilliant! I love it!

groovecoder commented:
I brought in https://github.com/mozilla/blurts-server/pull/486/commits/cca431b90b8c21b4a5323c189169279033f3c636 by @lesleyjanenorton which includes wrapping the handlebars/HTML strings with `fluentFormat`. So I think we're at the full i18n level now?

@stasm @flodolo I think @lesleyjanenorton had a question for how to pass args/variables into fluent strings.

lesleyjanenorton commented on hbs-helpers.js:
```diff
@@ -1,23 +1,51 @@
+  return LocaleUtils.fluentFormat(req.supportedLocales, fluentID, args);
+}
+
+function breachResponse(req, id, breachTitle, breachCount) {
```
@stasm @flodolo Trying unsuccessfully to pass breachTitle and breachCount to the app.ftl. Any thoughts on what is going wrong here?

lesleyjanenorton replied:
This is what comes back: <img width="1002" alt="screen shot 2018-10-11 at 3 56 43 pm" src="https://user-images.githubusercontent.com/22355127/46836907-f67da880-cd78-11e8-8a14-f6fb77e99d4b.png">

flodolo commented:
Note: you might need to add me as a contributor to the repository for CODEOWNERS to work

flodolo commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,35 @@
+-product-name = Firefox Monitor
+
+layout-support = Support
```
A little more, please? :) Localizers don't see (or read in most cases) code

```
# Support should be treated as a name, used in a link to ???
```

flodolo commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,304 @@
+
+pwt-section-headline = Passwords = Protection
+pwt-section-subhead = Your Private Information Is Only as Safe as Your Passwords
+pwt-section-blurb =  Your passwords protect more than your accounts, they protect every bit of personal information that resides in them. And hackers rely on bad habits, like using the same password everywhere or using common phrases (p@ssw0rd, anyone?), so that if they hack one account, they can hack many. Here are six ways to protect your accounts.
```
Extra leading space

flodolo commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,304 @@
+
+pwt-headline-1 = Use a Different Password<br />for Every Account
+pwt-headline-2 = Create Strong<br />Passwords
+pwt-headline-3 = Treat Security Questions<br>Like Extra Passwords
```
Would be nice to be consistent: `<br>` vs `<br />`

lesleyjanenorton replied:
I'm thinking these line breaks need to go since there isn't a way to know exactly where the line breaks need to be placed in each translation.

flodolo commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,304 @@
+featured-breach-only = Your email address appears in the <span class="featured-breach-title"> --- </span> breach, but does not appear in any other known breaches.
+
+featured-breach-results = Your email address appears in the { $featuredBreachTitle } breach, { $breachCount ->
+  [1] one other breach.
```
Do we really need to spell out numbers up to ten? Do we have data on the average number of breaches?

flodolo replied:
It [does](https://pontoon.mozilla.org/en-CA/firefox/browser/browser/preferences/preferences.ftl/?search=%5B1%5D&string=176248). But I'm almost sure the UI would be useful with this many variants, and they would be automatically exposed to all languages, instead of a standard plural form (we should verify it somehow).

flodolo commented:
One more note looking at the size of content from the last time I checked:
* We'll *need* a dev server, updated frequently, for localizers to test their translations.
* We'll need a system to keep partial localizations out of the production server, and only enable them after a certain percentage of completion.

flodolo commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,304 @@
+pwt-section-subhead = Your Private Information Is Only as Safe as Your Passwords
+pwt-section-blurb =  Your passwords protect more than your accounts, they protect every bit of personal information that resides in them. And hackers rely on bad habits, like using the same password everywhere or using common phrases (p@ssw0rd, anyone?), so that if they hack one account, they can hack many. Here are six ways to protect your accounts.
+
+pwt-headline-1 = Use a Different Password<br />for Every Account
```
This is going to backfire badly as soon as you need to change any of these strings, since you're generating IDs in the code later one. Please let's not do this.

```properties
 pwt-headline-strong-password = Use a Different Password<br />for Every Account 
 pwt-summary-strong-password = You can't prevent a data breach, but you can limit your exposure by always using different passwords for different websites. 
```

Then store the mapping somewhere

```
{
  1: 'strong-password',
  2: 'foo',
}
```

stasm replied:
Even better, use the full identifier (`pwt-headline-strong-password`) for extra-easy `grep` queries :)

lesleyjanenorton commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,304 @@
+  [9] nine other breaches.
+  [10] ten other breaches.
+  [other] { $breachCount } other breaches.
+}
```
Success- thank you! Working exactly as expected now after adding the * and indentation.

lesleyjanenorton commented on public/locales/en/app.ftl:
```diff
@@ -97,39 +100,35 @@ form-signup-checkbox = Get the latest from {-brand-Mozilla} and {-brand-name}.
+  That's good news, but data breaches can happen any time and there is still more you can do. 
+  Subscribe to Firefox Monitor for a full report, alerts when new breaches happen, and tips on protecting your passwords.
+
+featured-breach-results = 
```
@stasm I'm still running into some selector trouble here in "featured-breach-results". As is, this throws a "ctor is not a constructor" error. If I change [one] to [1], the string will return correctly only if the value of $breachCount is exactly 1. I'm totally puzzled as I recreated this over in the Fluent playground where it appears to be working exactly as expected. Have you run into anything like this?

stasm replied:
This might be related to the `Intl.PluralRules` API. It's still a fairly new feature in JavaScript. Looking at the [browser support matrix](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/PluralRules#Browser_compatibility), is there a chance you're testing on Safari? In order to support Safari as well as older browsers, you'll need to polyfill `Intl.PluralRules`. I recommend [`intl-pluralrules`](https://www.npmjs.com/package/intl-pluralrules).

Background: Fluent relies on this API to match variant keys which are not numbers. When you use `[one]` Fluent maps the numerical value of `$breachCount` to a plural category. Categories are [defined in CLDR](http://www.unicode.org/cldr/charts/29/supplemental/language_plural_rules.html). For English, they're `one` and `other`. Other languages have fewer or more.

groovecoder replied:
Aha, node.js is probably the culprit here - it doesn't have support for `Intl.PluralRules` API. I'll see if I can add that to our `package.json` and get this to work.

groovecoder commented on locale-utils.js:
```diff
@@ -0,0 +1,74 @@
+const path = require("path");
+
+// node.js needs Intl.PluralRules polyfill
+require("intl-pluralrules");
```
![image](https://user-images.githubusercontent.com/71928/46970260-38ab3080-d07e-11e8-974d-f5c94618eb48.png)

groovecoder commented on locale-utils.js:
```diff
@@ -0,0 +1,74 @@
+const path = require("path");
+
+// node.js needs Intl.PluralRules polyfill
+require("intl-pluralrules");
```
![its-working-its-working](https://user-images.githubusercontent.com/71928/46970288-4f518780-d07e-11e8-9721-c4ed5fbb3606.gif)

stasm replied:
Nice! Happy to hear that was it!

groovecoder commented:
Updated to have `fluentFormat` add messages from all `.ftl` files in the in `public/locales/{lang}/` directories.

lesleyjanenorton commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,185 @@
+    }.
+
+scan-results = Your account { $breachCount ->
+    [0] Your email address did not appear in our basic scan.
```
@flodolo thanks for the feedback. Is it possible to translate any variant in Pontoon, or is there a specific issue with the zero case?

lesleyjanenorton replied:
<img width="1363" alt="screen shot 2018-10-16 at 4 25 07 pm" src="https://user-images.githubusercontent.com/22355127/47048485-26092800-d160-11e8-97f6-d8c327e9f5f7.png">
Was following this example and a similar example of usage here: https://projectfluent.org/play/

Pike replied:
This is perfectly fine Fluent, it's just tricky for localizers to understand what we're asking them to translate.

It becomes easier if we're exposing the full phrase for each variant.

There's things we'll want to do on the l10n tooling side there in the next year, but right now, the extra repetition of the head and tail section would be good.

```properties
single-featured-breach-result = Your account appeared in the <span class="bold"> { $featuredBreach } </span> breach, but did not appear in any other known data breaches.
multiple-featured-breach-results =
  {  $breachCount ->
        [one] Your account appeared in the <span class="bold"> { $featuredBreach } </span> breach, as well as { $breachCount } other breach.
       *[other] Your account appeared in the <span class="bold"> { $featuredBreach } </span> breach, as well as { $breachCount } other breaches.
  }
```

lesleyjanenorton replied:
@Pike thanks for the clarification

flodolo replied:
What Pike said :)

The UI looks [like this](https://pontoon.mozilla.org/it/firefox/browser/browser/preferences/preferences.ftl/?search=%5B1%5D&string=176248). The fragment before would become another input field, and you would need to drag the input field to see everything.

flodolo replied:
To be clear, I would do this

```
featured-breach-no-results = 
  Your account appeared in the <span class="bold"> { $featuredBreach } </span> breach, 
  but did not appear in any other known data breaches.

featured-breach-results =
    { $breachCount ->
        [one] Your account appeared in the <span class="bold"> { $featuredBreach } </span> breach, as well as { $breachCount } other breach.
      *[other] Your account appeared in the <span class="bold"> { $featuredBreach } </span> breach, as well as { $breachCount } other breaches.
    }

scan-no-results = 
    Your email address did not appear in our basic scan.
    That's good news, but data breaches can happen any time and there is still more you can do. 
    Subscribe to Firefox Monitor for a full report, alerts when new breaches happen, and tips on protecting your passwords.
    
scan-results = { $breachCount ->
    [one] Your account appeared in the following breach.
  *[other] Your account appeared in the following { $breachCount } breaches.
}
```

Note also that your current string would result in `Your account Your email address did not appear in our basic scan. etc`

stasm replied:
> I really think we should split out the 0 case. If nothing else for practical reasons: it would be impossible to translate this in Pontoon, where there no text area for variants.

From the Fluent perspective, I think using `[0]` like this is fine and it adds flavor to the translation. The [Good Practices document](https://github.com/projectfluent/fluent/wiki/Good-Practices-for-Developers#prefer-separate-messages-over-variants-for-ui-logic) has more discussion about this.

However, if Pontoon doesn't support extra variants right now, we should avoid it (for now). Sorry, @lesleyjanenorton! We'll get there eventually :)

lesleyjanenorton commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,185 @@
+# Descriptive headline for a column of links where users can give feedback, or get additional information about, Firefox Monitor.
+layout-support = Support
+# Link that takes the user to a Firefox Monitor survey. 
+give-feedback = <span class="nowrap">Give Feedback</span>
```
Yep, we can move this to template.

lesleyjanenorton replied:
I'll remove it wherever possible. Unfortunately, for cases like L82 and a few others, it will need to remain as is.

lesleyjanenorton commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,185 @@
+  And hackers rely on bad habits, like using the same password everywhere or using common phrases (p@ssw0rd, anyone?), 
+  so that if they hack one account, they can hack many. Here are six ways to protect your accounts.
+
+pwt-headline-1 = Use a Different Password for Every Account
```
I think in this case the numbers are meaningful because the password tips are specifically presented in the order that they are numbered.  Additionally, the tips are displayed alongside corresponding icons which are also numbered accordingly.  I will happily defer to your judgement here if you think this is going to cause future problems though.

lesleyjanenorton replied:
Regarding the colon spacing on L13 (for some reason I can't respond to that comment directly): We just received updated copy for the emails and the extra space has been removed.

flodolo replied:
> Additionally, the tips are displayed alongside corresponding icons which are also numbered accordingly.

How is that done? Have you considered other types of numerals?

>  I will happily defer to your judgement here if you think this is going to cause future problems though.

Two likely scenarios:
* You want to change the order. 
* You want to replace or update one of the tips. You'll need to [change the message ID](https://developer.mozilla.org/en-US/docs/Mozilla/Localization/Localization_content_best_practices#Changing_existing_strings) to do that.

In both cases you'll need to add a mapping afterward.

Also, strings are presented sequentially to localizers, so having title and description together help a bit.

lesleyjanenorton commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,185 @@
+-brand-Quantum = Firefox Quantum
+-brand-Mozilla = Mozilla
+-brand-HIBP = HIBP
+-product-name-nowrap = <span class="nowrap">Firefox Monitor</span>
```
Potentially - we've discussed this in previous pull requests and my understanding is that &nbsp; is deprecated and there are a11y concerns with using it.

lesleyjanenorton commented on public/locales/en/app.ftl:
```diff
@@ -0,0 +1,185 @@
+error-not-subscribed = This email address is not subscribed to {-product-name}.
+error-hibp-throttled = Too many connections to {-brand-HIBP}.
+error-hibp-connect = Error connecting to {-brand-HIBP}.
+error-hibp-load-breaches = Could not load breaches.
```
I think L16-L18 are actually a log errors and may not need to be localized at all.

flodolo commented on public/locales/en/data-classes.ftl:
```diff
@@ -0,0 +1,117 @@
+email-messages = Email messages
+employers = Employers
+ethnicities = Ethnicities
+family-members'-names = Family members' names
```
I'm pretty sure the `'` in the ID is a parsing error. The string, on the other hand, should use an apostrophe `’`

groovecoder commented:
After @lesleyjanenorton's updates, I added commits to remove the test `de/app.ftl` file, to move `public/locales` to `locales` at the base of the repo, and to check for `NODE_ENV === "heroku"` to enable pontoon JS for in-page localization.

I've also pushed this branch directly to https://fx-breach-alerts.herokuapp.com/ so we can see this code running live before our upcoming meeting.

(Note: @lesleyjanenorton I'm getting `error {"stack":"TypeError: dataClasses.split is not a function
    at Object.breachDataClasses (/Users/lcrouch/code/mozilla/blurts-server/hbs-helpers.js:28:71)` on http://localhost:6060/user/verify?token=87dea900-428d-4b71-aad9-16b9ceb7ed92. Trying to track it down now.)

flodolo commented:
>  to check for `NODE_ENV === "heroku"` to enable pontoon JS for in-page localization.

I'm not exactly sure if that works reliably, or for Fluent projects. @mathjazz ?

mathjazz commented:
> I'm not exactly sure if that works reliably, or for Fluent projects. @mathjazz ?

In theory it works, but requires additional testing. AFAICT, it hasn't been used in combination with a Fluent.

I suggest we don't block enabling localization on it and enable in-place localization as a follow-up.

groovecoder commented:
In our case, I've specifically set `NODE_ENV=heroku` using `heroku:config` so it should work for us. Not sure what the broader `NODE_ENV` convention is.

I've also fixed the error on `/user/verify` so the text actually shows, though the page styles are messed up.

![screen shot 2018-10-17 at 11 17 44 am](https://user-images.githubusercontent.com/71928/47101005-8353b680-d1fe-11e8-91db-b403d6344b3b.png)

groovecoder commented on locales/en/app.ftl:
```diff
@@ -0,0 +1,193 @@
+# - Declined to adapt to grammatical case.
+# - Transliterated.
+# - Translated.
+-product-name = Firefox Monitor
```
Filed https://github.com/mozilla/blurts-server/issues/535 for a follow-up on this.

groovecoder commented:
I'm merging this now, as it touches so many files and is preventing us from doing much else until it's merged.

Filed https://github.com/mozilla/blurts-server/issues/533 for email localization, https://github.com/mozilla/blurts-server/issues/534 for the heroku/in-page localization follow-up, https://github.com/mozilla/blurts-server/issues/535 for `&nbsp;` discussion & decision, and  https://github.com/mozilla/blurts-server/issues/524 to verify that @mozilla-pontoon can/will create the launch `.ftl` files in the repository for launch.

They are all in [the `l10n` label](https://github.com/mozilla/blurts-server/issues?q=is%3Aissue+is%3Aopen+label%3Al10n) now.
